xde/src/xde.rs

```diff
@@ -1810,78 +2094,567 @@ fn guest_loopback(
+                        xde.stats.vals.mcast_tx_stale_local().incr(1);
                     }
                 }
+            }
+        }
+    }
+
+    // Next hop forwarding: send packets to configured next hops.
+    //
+    // At the leaf level, we process all next hops in the forwarding table.
+    // Each next hop's `Replication` is a TX-only instruction telling the switch
+    // which ports to replicate to:
+    // - External: ports set for external multicast traffic (egress to external networks)
+    // - Underlay: replicate to other sleds (using multicast outer dst)
+    // - Both: both external and underlay replication
+    //
+    // We already have the Arc from the per-CPU cache, no need to clone.
+    if cpu_mcast_fwd.get(&ctx.underlay_dst).is_none() {
+        __dtrace_probe_mcast__no__fwd__entry(
+            &ctx.underlay_dst,
+            ctx.vni.as_u32() as uintptr_t,
+        );
+        let xde = get_xde_state();
+        xde.stats.vals.mcast_tx_no_fwd_entry().incr(1);
+    }
+
+    if let Some(next_hops) = cpu_mcast_fwd.get(&ctx.underlay_dst) {
+        // We found forwarding entries, replicate to each next hop
+        for (next_hop, replication) in next_hops.iter() {
+            // Clone packet with headers using pullup
+            let Ok(mut fwd_pkt) =
+                ctx.out_pkt.pullup(NonZeroUsize::new(pullup_len))
+            else {
+                opte::engine::dbg!(
+                    "mcast TX next hop pullup failed: requested {} bytes",
+                    pullup_len
+                );
+                let xde = get_xde_state();
+                xde.stats.vals.mcast_tx_pullup_fail().incr(1);
+                __dtrace_probe_mcast__tx__pullup__fail(pullup_len as uintptr_t);
+                continue; // Skip this destination on allocation failure
+            };
+
+            // Route to next hop unicast address to determine which underlay
+            // port/MAC to use. Packet destination is the multicast address with
+            // multicast MAC (RFC 2464).
+            //
+            // NextHopV6.addr = unicast switch address (for routing)
+            // Outer dst IP = ctx.underlay_dst (multicast address from M2P)
+            // Geneve Replication is a TX-only instruction telling the switch
+            // which port groups to use.
+            let routing_dst = next_hop.addr;
+            let actual_outer_dst = ctx.underlay_dst;
+
+            // VNI is at offset 4 in Geneve header (3 bytes)
+            if let Some(vni_bytes) =
+                fwd_pkt.get_mut(geneve_offset + 4..geneve_offset + 7)
+            {
+                let vni_be = next_hop.vni.as_u32().to_be_bytes();
+                vni_bytes.copy_from_slice(&vni_be[1..4]); // VNI is 24 bits
+            }
+            // Update Geneve multicast option to reflect underlay replication to
+            // prevent re-relay loops.
+            update_mcast_replication(&mut fwd_pkt, geneve_offset, *replication);
```

Comment:
Nit: dendrite should already be preventing reflection/re-relay, I think. IIRC we have the logic for it not to mcast-forward on the port a packet came inbound on (or, at least, to drop during egress processing).